extractor.go

```diff
@@ -186,87 +188,112 @@ func (e *Extractor) Tar(ctx context.Context, body io.Reader, location string, re
-		// filename... oh, well...
-		forceDir := strings.HasSuffix(path, "\\")
-		path = strings.Replace(path, "\\", "/", -1)
+			// Replace backslash with forward slash. There are archives in the wild made with
+			// buggy compressors that use backslash as path separator. The ZIP format explicitly
+			// denies the use of "\" so we just replace it with slash "/".
+			// Moreover it seems that folders are stored as "files" but with a final "\" in the
+			// filename... oh, well...
+			forceDir := strings.HasSuffix(path, "\\")
+			path = strings.Replace(path, "\\", "/", -1)
 
-		if rename != nil {
-			path = rename(path)
-		}
+			if rename != nil {
+				path = rename(path)
+			}
 
-		if path == "" {
-			continue
-		}
+			if path == "" {
+				continue
+			}
 
-		if path, err = safeJoin(location, path); err != nil {
-			continue
+			var err error
+			if path, err = safeJoin(location, path); err != nil {
+				continue
+			}
+
+			info := header.FileInfo()
+
+			switch {
+			case info.IsDir() || forceDir:
+				if err := e.FS.MkdirAll(path, info.Mode()|os.ModeDir|100); err != nil {
+					return errors.Annotatef(err, "Create directory %s", path)
+				}
+			// We only check for symlinks because hard links aren't possible
+			case info.Mode()&os.ModeSymlink != 0:
+				if f, err := header.Open(); err != nil {
+					return errors.Annotatef(err, "Open link %s", path)
+				} else if name, err := ioutil.ReadAll(f); err != nil {
+					return errors.Annotatef(err, "Read address of link %s", path)
+				} else {
+					links = append(links, link{Path: path, Name: string(name)})
+					f.Close()
+				}
+			default:
+				if f, err := header.Open(); err != nil {
+					return errors.Annotatef(err, "Open file %s", path)
+				} else if err := e.copy(ctx, path, info.Mode(), f); err != nil {
+					return errors.Annotatef(err, "Create file %s", path)
+				} else {
+					f.Close()
+				}
+			}
 		}
+		return nil
+	}
 
-		info := header.FileInfo()
+	for i := 0; i < runtime.NumCPU(); i++ {
+		g.Go(filesWorker)
+	}
+	if err := g.Wait(); err != nil {
+		return err
+	}
 
-		switch {
-		case info.IsDir() || forceDir:
-			if err := e.FS.MkdirAll(path, info.Mode()|os.ModeDir|100); err != nil {
-				return errors.Annotatef(err, "Create directory %s", path)
-			}
-		// We only check for symlinks because hard links aren't possible
-		case info.Mode()&os.ModeSymlink != 0:
-			if f, err := header.Open(); err != nil {
-				return errors.Annotatef(err, "Open link %s", path)
-			} else if name, err := ioutil.ReadAll(f); err != nil {
-				return errors.Annotatef(err, "Read address of link %s", path)
-			} else {
-				links = append(links, link{Path: path, Name: string(name)})
-				f.Close()
+	// Processes links, this is done in a second moment since it might happen that
+	// the file to link still has not been exctracted, that would make the linking fail
+	g, ctx = errgroup.WithContext(originalCtx)
+	for _, link := range links {
+		g.Go(func() error {
```

Comment:
This loop starts `len(links)` goroutines at the same time, this may be not what we want :-)

Reply:
Yup, that's a leftover from a different approach.